### PR TITLE
Add alias field to rules

### PR DESCRIPTION
**Pull request by rgrinberg**
### This PR

- [x] `rule` supports `alias` from `2.0`
- [x] rules no longer need a set of targets if there's an `alias` field.

### Upcoming PR's

- [ ] Using `action` in `alias` should be deprecated in 1.x. This deprecation warning will be quite annoying given that we have no way to silence it.
- [ ] Disallow using `action` in alias entirely for >= 2.0.
- [x] Allow for `pacakge` on rules.

### CODE REVIEW

**Comment by aalekseyev** [comment]
This interpretation is correct in principle, but I'm not sure about the details.
If you have a rule that declares some targets and an alias, won't there be a conflict between the two copies of the action? (one in the rule, another in the alias) (at the very least it might be run twice, but at worst it might overwrite the targets when the build system doesn't expect)

**Comment by rgrinberg** [comment]
Yeah, it definitely seems problematic. In fact, this attempt fails even earlier because it's not possible the rule action to update the alias and the rule's targets. Those targets exist in different directories (as the test shows).

**Comment by rgrinberg** [comment]
Now I'm wondering if there's a way to implement without some roundabout hacks. For example, it seems like if we could add some dummy target to the rule and make it a  dependency of the alias that would work fine. But then we'd have to make sure this dummy target is hidden from that user. This hack doesn't seem desirable to me.

**Comment by diml** [comment]
Isn't using a dummy target what we do now for aliases with action?

**Comment by rgrinberg** [comment]
@aalekseyev have a look at my last commit. Do you think this is more reasonable?

My only issue is that this dummy file can be accidentally caught through globs :/

**Comment by aalekseyev** [comment]
I haven't looked at the code yet, but the discussion in the RFC made me think: maybe we should start by simply translating the rule definitions to alias definitions internally when they come without targets.

**Comment by rgrinberg** [comment]
I take it you mean `rule` stanzas without targets but with an `alias` field. Sure we can start with that, but we'll still need to support the case with both `alias` and target so it's not like we can do one without the other.

**Comment by rgrinberg** [comment]
This PR now also supports creating rules without targets and just an `alias` field.

**Comment by rgrinberg** [comment]
I just discovered a use case that cannot be replaced with adding `alias` to `rule`. With the `alias` stanza it's possible to set the `package` field. This is useful for attaching tests to particular packages. I suppose we need something like this for `rule` as well.

**Comment by aalekseyev** [comment]
To be honest, I don't really understand packages. I guess a package, from dune point of view, is:
1. a certain set of build targets that you can build (e.g. with -p)
2. a certain set of files that the aforementioned set of targets is allowed to depend on

I'm not sure which of these we want to interact with and how. My guess is it's (1): we just want to make the build of a given package depend on a given alias.

If so, maybe the right way to think about it is that there's a special alias associated with each package. (let's say its name is (package foo))

If you want a rule to be built as a part of a package, you write (rule ... (alias (package blah))) or (rule ... (package blah)) as an abbreviation. (only allowing the second syntax only is fine too, but I think desugaring in terms of aliases would at least clarify the semantics).

In fact, it might be a good idea to allow a rule to register itself with multiple aliases.

**Comment by rgrinberg** [comment]
It’s not yet clear to me how your idea would work with tests however. Without filtering the rules with -p, how do you propose we run tests for an individual package?

I still like your idea because I think it makes sense to move away from relying on -p for running package tests. I would be much more comfort with tests for a particular package being selected as just a normal build target.

By the way this PR is independent of these considerations as the old alias stanza still works.

File: test/blackbox-tests/test-cases/github2681/run.t [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  $ cat > dune <<EOF
+  > (rule
+  >  (action (with-stdout-to foo (echo "hello world")))
+  >  (alias bar))
+  > EOF
+  $ dune build @bar --display short
+  $ cat _build/default/foo
+  hello world
+
+The field order should not affect the alias field:
```
Comment (aalekseyev) [inline test/blackbox-tests/test-cases/github2681/run.t]: Is this really worth testing? Isn't the field parser already making sure of that?

File: src/dune/simple_rules.ml [hunk 2]
```diff
@@ -35,6 +35,16 @@ let check_filename =
...
 
+let add_alias sctx ~dir ~name ~stamp ~loc ?(locks = []) build =
+  let alias = Alias.make name ~dir in
+  SC.add_alias_action sctx alias ~dir ~loc ~locks ~stamp build
+
+let stamp ~deps ~action ~extra_bindings =
+  ( "user-alias"
+  , Bindings.map ~f:Dep_conf.remove_locs deps
+  , Option.map ~f:Action_unexpanded.remove_locs action
+  , Option.map extra_bindings ~f:Pform.Map.to_stamp )
```
Comment (aalekseyev) [inline src/dune/simple_rules.ml]: While we're at this, do you know what `extra_bindings` are and why they're used here? I know it's a pre-existing thing, but it seems weird to me.

File: src/dune/simple_rules.ml [hunk 3]
```diff
@@ -57,6 +67,29 @@ let user_rule sctx ?extra_bindings ~dir ~expander (rule : Rule.t) =
...
+      match rule.alias with
+      | None -> action
+      | Some alias ->
+        let stamp =
+          let action = Some (snd rule.action) in
+          stamp ~deps:rule.deps ~extra_bindings ~action
+          |> Digest.generic
+          |> Digest.to_string
+        in
+        let stamp_target = Path.Build.relative dir stamp in
```
Comment (aalekseyev) [inline src/dune/simple_rules.ml]: I expect a stamp target matching random globs to be pretty annoying. We should probably do something about that.

One (somewhat hacky) approach is, as I said before, to move the rule action to the alias dir if there's no targets. (I think alias actions are translated to rules internally in builds_system.ml anyway so it's perhaps not as much of a hack as it seems at first)

Another thing we could do is to allow targets in multiple directories as long as the set of directories is known in advance, but that seems trickier.

File: src/dune/simple_rules.ml [hunk 4]
```diff
@@ -57,6 +67,29 @@ let user_rule sctx ?extra_bindings ~dir ~expander (rule : Rule.t) =
...
+        let () =
+          let alias = Alias.make alias ~dir in
+          let deps = Path.Set.singleton (Path.build stamp_target) in
+          Rules.Produce.Alias.add_deps alias deps
```
Comment (aalekseyev) [inline src/dune/simple_rules.ml]: I guess `add_deps` currently adds the dep both as a "transitive dep" and as a a part of the alias expansion? It seems wrong to add the stamp to alias expansion, but if there's no away to avoid that it's OK to leave it as is (and add a comment, perhaps).

File: test/blackbox-tests/test-cases/github2681/run.t (hunk 1 above)
Comment (aalekseyev) [reply test/blackbox-tests/test-cases/github2681/run.t]: Could you add a test with non-empty set of targets as well (maybe instead of the reordering test)?

**Review by aalekseyev — DISMISSED** [review]
Things look good other than polluting globs with the stamp file names. I think we should probably fix that.

**Comment by aalekseyev** [comment]
Regarding packages, I'm happy to leave it for later. Thanks for pointing out my ideas don't match the reality. I'll need to understand the existing thing at some point though.

File: test/blackbox-tests/test-cases/github2681/run.t (hunk 1 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/github2681/run.t]: I removed the extra test. In fact, my parsing code was a little wrong before so this failed.

I assume you meant an empty set of targets. I added such a test.

File: src/dune/simple_rules.ml (hunk 2 above)
Comment (rgrinberg) [reply src/dune/simple_rules.ml]: One example where it's used is to introduce the `%{test}` binding for the `test` stanza.

File: src/dune/simple_rules.ml (hunk 3 above)
Comment (rgrinberg) [reply src/dune/simple_rules.ml]: > One (somewhat hacky) approach is, as I said before, to move the rule action to the alias dir if there's no targets. (I think alias actions are translated to rules internally in builds_system.ml anyway so it's perhaps not as much of a hack as it seems at first)

But what about rules with targets? We'd be left with the same problem in this case.

> Another thing we could do is to allow targets in multiple directories as long as the set of directories is known in advance, but that seems trickier.

This is highly desirable. It's required to support  other kinds of rules such as coq's native compute rules.

However, I have a simpler solution for this problem. What I think we need is a "private" notion of targets. Such private targets:

* Cannot be picked up by user written globs
* Cannot be used as explicit dependencies for user written rules.

Such targets would solve other similar problems:
* Users relying on implementation specific paths like `.objs` and then dune changes end up breaking them.
* They make it easy for us to exclude undesired stuff from `@all`.

File: src/dune/simple_rules.ml (hunk 4 above)
Comment (rgrinberg) [reply src/dune/simple_rules.ml]: I've added a TODO for now. Indeed I don't see any other way.

File: src/dune/simple_rules.ml (hunk 3 above)
Comment (aalekseyev) [reply src/dune/simple_rules.ml]: > But what about rules with targets?

For rules with targets we don't really need a stamp file because we can depend on any of those targets.

> This is highly desirable.

Yes indeed!

> "private" notion of targets

That idea only helps if the rules are sandboxed. If they are not sandboxed, they will see the entire directory and dune has no power over that. Imagine somebody writing a rule "cat *" and declaring that it depends on glob "*".

A safe solution is to make it an error when a glob for a non-sandboxed rule happens to include a private dep, but that's obviously not great.

**Comment by rgrinberg** [comment]
> Yes indeed!
I suppose we should open a separate ticket for this.  Do you have an implementation in mind?
> That idea only helps if the rules are sandboxed. If they are not sandboxed, they will see the entire directory and dune has no power over that. Imagine somebody writing a rule "cat " and declaring that it depends on glob "".
yes that’s a good point. Okay you’ve convinced me that this would be too fragile. I suppose the priority for multi dir rules is fairly high now. I’ll try to remember to mention it at the next dune meeting.

**Comment by rgrinberg** [comment]
Another thing that is worth mentioning here is that the locks field is supported by the alias stanza but not the rule stanza. I will make an appropriate pr for that as there’s no reason for locks not work on normal rules

**Comment by rgrinberg** [comment]
I've added package to `rule` in this PR as well. I suppose there's no way to avoid it.

**Comment by diml** [comment]
Just a note: we should aim at having a uniform handling for all rules that are part of an alias, whether or not they have targets, and we should also try to find a way to not rely on these stamp files that tend to get in the way. Maybe we need some kind of virtual targets?

In any case, I'm removing this from the 2.0.0 milestone. Looking back at it, it feels like a too big change to go through just before the release.

**Comment by diml** [comment]
> I've added package to rule in this PR as well. I suppose there's no way to avoid it.

That seems right to me.

**Comment by rgrinberg** [comment]
@diml @aalekseyev this PR is ready. I made this feature available for 2.1.

**Comment by aalekseyev** [comment]
I wasn't sure how to run your tests so I pushed some ridiculous hack in `79f2639d3ddfc79e2cc187569475f24b9d80178c`. What do you think about the hack? Maybe it's not so ridiculous after all?

**Comment by rgrinberg** [comment]
Seems like a useful hack to me. We could use this thing in the ocamlfdo PR as well perhaps.

File: src/dune/simple_rules.ml [hunk 5]
```diff
@@ -35,10 +56,43 @@ let check_filename =
...
+    ~loc:rule.loc
+    ~locks:(interpret_locks ~expander rule.locks)
+    action
+
 let user_rule sctx ?extra_bindings ~dir ~expander (rule : Rule.t) =
   match Expander.eval_blang expander rule.enabled_if with
-  | false -> Path.Build.Set.empty
-  | true ->
+  | false ->
+    Option.iter rule.alias ~f:(fun name ->
```
Comment (aalekseyev) [inline src/dune/simple_rules.ml]: What's the idea here? Why are we adding something at all if the rule is disabled?

File: src/dune/simple_rules.ml [hunk 6]
```diff
@@ -35,10 +56,43 @@ let check_filename =
...
+type rule_kind =
+  | Alias_only of Alias.Name.t
+  | Alias_with_targets of Alias.Name.t * Path.Build.t
+  | No_alias
+
+let rule_kind ~(rule : Rule.t) ~action =
+  match rule.alias with
+  | None -> No_alias
+  | Some alias -> (
+    match Build.targets action |> Path.Build.Set.choose with
```
Comment (aalekseyev) [inline src/dune/simple_rules.ml]: Is it well-defined what choice `choose` makes? I think we should consistently make the same choice.

File: src/dune/simple_rules.ml (hunk 5 above)
Comment (rgrinberg) [reply src/dune/simple_rules.ml]: The idea is that we want `$ dune build @foo` to work regardless if the rule is enabled or not. Consider `$ dune build @runtest` where a test might potentially be disabled. We don't want to get an error like "this alias is empty"

File: src/dune/simple_rules.ml (hunk 6 above)
Comment (rgrinberg) [reply src/dune/simple_rules.ml]: True, but I can't think of another way to consistently make the same choice. Do you have a suggestion?

File: src/dune/simple_rules.ml (hunk 5 above)
Comment (aalekseyev) [reply src/dune/simple_rules.ml]: Ok, fair enough, but you're still passing rule.deps. Do you mean empty deps?

Comment (rgrinberg) [reply src/dune/simple_rules.ml]: Hmm, good point. I'm not sure why I did it that way but it's wrong. The deps are no longer interpreted.

File: src/dune/simple_rules.ml (hunk 6 above)
Comment (aalekseyev) [reply src/dune/simple_rules.ml]: If the answer is "yes" then I don't have a better suggestion, this is good. If the answer is no, then we should choose (e.g.) the minimum instead.

Comment (aalekseyev) [reply src/dune/simple_rules.ml]: Looks like the answer is "yes", at least in the sense that matters:

```
    (** Return one element of the given set, or raise [Not_found] if
       the set is empty. Which element is chosen is unspecified,
       but equal elements will be chosen for equal sets. *)
```

**Comment by aalekseyev** [comment]
This looks good to me.

File: src/dune_lang/syntax.ml [hunk 7]
```diff
@@ -43,6 +43,8 @@ module Version = struct
     parser_major = data_major && parser_minor >= data_minor
 end
 
+let inside_dune = ref false
```
Comment (diml) [inline src/dune_lang/syntax.ml]: This doesn't look right to me. It doesn't impact only dune's own tests as other users sometimes also run dune inside dune. We could add a special version for such cases, for instance `(lang dune 2.1+dev)`. In this particular case, we might as well put the feature back in 2.0.0.

**Comment by rgrinberg** [comment]
Shall we have a test mode perhaps? I think this is still a useful feature to have.

**Comment by diml** [comment]
A test mode seems fine, as long as it is of low risk of being enabled accidentally by the user. And if it's enabled via an environment variable, there is no need to rely on a reference.

Regarding this particular feature, just thinking out loud: I believe we agreed that in this case:

```scheme
(rule
 (deps x)
 (targets y)
 (alias foo)
 (action (run blah)))
```

then none of `x` or `y` are direct dependency of alias `foo`. However, this PR adds `y` as a dependency of `foo`. Is that correct? In this case, we should make sure to change that before this feature is made available.

Another thought: once this is all finished, I don't believe there is anymore a need for `Expander.Targets.Forbidden`. And maybe `Infer` as well?

**Comment by rgrinberg** [comment]
> then none of x or y are direct dependency of alias foo. However, this PR adds y as a dependency of foo. Is that correct? In this case, we should make sure to change that before this feature is made available.
Yes, that is indeed the case. However, I’m not sure how to do this without introducing the stamp file and that has its own caveats. I believe that we should be able to address this issue once we have a proper alias database. Do you think making this change later will break user code down the line? I thought that changing this would be unobservable to user rules.

**Comment by diml** [comment]
Yh indeed. Well, it might be possible to observe a difference in terms of incremental compilation, but it's probably obscure enough to go completely unnoticed.

Do we have any more reason to not have the new fields in 2.0.0 then?

**Comment by rgrinberg** [comment]
> Do we have any more reason to not have the new fields in 2.0.0 then?
No objection for allowing it in 2.0. I’ll make a PR.

**Comment by aalekseyev** [comment]
> inside_dune

@diml is making a good point. I'll revert it, but I think we should probably make an alternative at some point.

> [alias expansion details]

Incremental compilation is not the only way to observe the differences: if you're building with sandboxing then the contents of the sandbox will also reflect the deps.
That said, I think it's fine to leave it for now: the stamp file itself is equally undesirable to have in the expansion, and the cleanup in this area could be done as an independent improvement.

**Comment by aalekseyev** [comment]
I pushed the revert to master, and only then I realized that I broke tests, after which the simplest fix was to move the feature to 2.0.0, which I pushed to master too: 84b7fde99d8913b71be3a1a21ccc562e0251d528.
